samples: sensor: Merge magnetometer samples

MaureenHelm opened the pull request:
Merges several driver-specific magnetometer samples into one more generic magnetometer sample.

MaureenHelm committed:
drivers: sensor: Use dt_compat_enabled for magnetometer drivers

In preparation for extending samples/sensor/magn_polling to support
additional magnetometer drivers, enable these drivers by default when
the sensor driver class is enabled (CONFIG_SENSOR=y) and a compatible
devicetree node is enabled.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
samples: sensor: Remove GPIO references from magn_polling

The magn_polling sample application doesn't use GPIO.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
samples: sensor: Extend magn_polling to support additional drivers

Extends the magn_polling sample application to support additional
magnetometer drivers. This is a first step towards reducing the number
of driver-specific sensor sample applications.

Integration platforms were selected to exercise different magnetometer
drivers.

Tested on hardware with the following configurations:
- fxos8700 driver with the on-board sensor on the frdm_k64f board
- lis2mdl driver with a x_nucleo_iks01a3 shield attached to a frdm_k64f
  board

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
samples: sensor: Remove bmm150 magnetometer sample

The magn_polling sample application now supports the bmm150 driver, so
we can remove this driver-specific sample.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
samples: sensor: Remove hmc5883l magnetometer sample

The magn_polling sample application now supports the hmc5883l driver, so
we can remove this driver-specific sample.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

gmarull commented on samples/sensor/magn_polling/sample.yaml:
```diff
@@ -1,8 +1,26 @@
-    depends_on: i2c
+  sample.sensor.magn_polling.bmc150_magn:
     filter: dt_compat_enabled("bosch,bmc150_magn")
+    extra_args: MAGN_POLLING_COMPAT=bosch_bmc150_magn
+  sample.sensor.magn_polling.bmm150:
+    filter: dt_compat_enabled("bosch,bmm150")
+    extra_args: MAGN_POLLING_COMPAT=bosch_bmm150
+  sample.sensor.magn_polling.fxos8700:
+    filter: dt_compat_enabled("nxp,fxos8700")
+    extra_args: MAGN_POLLING_COMPAT=nxp_fxos8700
+    extra_configs:
+      - CONFIG_FXOS8700_MODE_HYBRID=y
+  sample.sensor.magn_polling.hmc5883l:
+    filter: dt_compat_enabled("honeywell,hmc5883l")
+    extra_args: MAGN_POLLING_COMPAT=honeywell_hmc5883l
+  sample.sensor.magn_polling.lis2mdl:
+    filter: dt_compat_enabled("st,lis2mdl")
+    extra_args: MAGN_POLLING_COMPAT=st_lis2mdl
+  sample.sensor.magn_polling.lsm303dlhc-magn:
+    filter: dt_compat_enabled("st,lsm303dlhc-magn")
+    extra_args: MAGN_POLLING_COMPAT=st_lsm303dlhc_magn
```
note: wouldn't it be a better idea to rely on aliases like we do for LEDs, buttons, etc.? Something like

```
aliases {
    magn-0 = &mymagnetometer;
};
```

then in the sample do

```
... DEVICE_DT_GET(DT_ALIAS(magn_0));
```

MaureenHelm replied:
My initial thought was that I might want to try to support multiple driver instances and have the application read all of the available sensors, but that might overcomplicate the sample. I gave your alias idea a try and I like how it turned out. Please have look at the new version.

MaureenHelm committed:
boards: Set magnetometer devicetree alias for fxos8700 nodes

Sets a magnetometer devicetree alias for all boards that contain the
FXOS8700 sensor to enable the magn_polling sample application.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
boards: Set magnetometer devicetree alias for bmm150 nodes

Sets a magnetometer devicetree alias for all boards that contain the
BMM150 sensor to enable the magn_polling sample application.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
boards: Set magnetometer devicetree alias for lis2mdl nodes

Sets a magnetometer devicetree alias for all boards that contain the
LIS2MDL sensor to enable the magn_polling sample application.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
boards: Set magnetometer devicetree alias for lsm303dlhc-magn nodes

Sets a magnetometer devicetree alias for all boards that contain the
LSM303DLHC sensor to enable the magn_polling sample application.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
boards: Set magnetometer devicetree alias for lsm303agr-magn nodes

Sets a magnetometer devicetree alias for all boards that contain the
LSM303AGR sensor to enable the magn_polling sample application.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
drivers: sensor: Change fxos8700 default mode from accel-only to hybrid

The fxos8700 device supports 3 modes: accelerometer-only,
magnetometer-only, or hybrid (accelerometer and magnetometer) modes. The
accelerometer-only mode is register compatible with mma8451q, mma8652fc,
and mma8653fc, which allows the fxos8700 driver to be used with these
devices as well.

Most in-tree boards can use hybrid mode because they have the fxos8700
device, therefore we change the driver default to match the common case.
For the handful of boards that have an mma86xx or mma84xx device, we
override the driver default to accelerometer-only mode. As a result, we
can enable the magn_polling sample application for the fxos8700 driver
without having to add a bunch of board-specific configuration overlays
for hybrid mode.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm commented:
> I also think we should probably add some guidelines on sensor samples to prevent more _duplicates_ in the future

Agreed, but I'd like to do that in a follow up PR.

gmarull reviewed (dismissed):
Nice cleanup, I think the alias approach will scale better for future sensors

gmarull commented on samples/sensor/magn_polling/sample.yaml:
```diff
@@ -1,8 +1,18 @@
+      - thingy53_nrf5340_cpuapp # bmm150
+      - stm32f411e_disco        # lsm303agr_magn
+      - stm32f3_disco           # lsm303dlhc_magn
+  sample.sensor.magn_polling.spi:
+    extra_configs:
+      -  CONFIG_SPI=y
+    integration_platforms:
+      - sensortile_box          # lis2mdl
```
I wonder if this can be handled in the board Kconfig.defconfig:

```
config SPI
    default y if XXX
```

erwango replied:
This would need to be done in each board on which we want to run this sample variant while proposed way allows to run the test with multiple boards.
So I think there pros and cons in both approaches.

gmarull replied:
this technique is used e.g. for the display sample, I find it useful because board specifics are kept in board directory. but not a blocker because we have no rules for that.

gmarull commented on samples/sensor/magn_polling/sample.yaml:
```diff
@@ -1,8 +1,18 @@
+      - thingy53_nrf5340_cpuapp # bmm150
+      - stm32f411e_disco        # lsm303agr_magn
+      - stm32f3_disco           # lsm303dlhc_magn
+  sample.sensor.magn_polling.spi:
+    extra_configs:
+      -  CONFIG_SPI=y
+    integration_platforms:
+      - sensortile_box          # lis2mdl
```
btw, prj.conf contains `CONFIG_I2C=y`, which will apply to this case as well (not desired)

erwango reviewed (dismissed):
Nice!

jfischer-no reviewed (dismissed):
:+1:

nashif reviewed (dismissed):
Great! I needed to revive my temp. sensort unification PR now :)

teburd reviewed (dismissed):
Nice!

MaureenHelm committed:
boards: Configure I2C/SPI default for boards with magnetometer alias

For all boards that have a magnetometer devicetree alias, configure the
appropriate bus (I2C or SPI) default to be enabled when the sensor
driver class is enabled. This will simplify enabling the magn_polling
sample application for these boards (and using the magnetometer in
general) because it will eliminate the need to add a bunch of
board-specific configuration overlays to the application.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
drivers: gpio: Change DW driver to use depends on instead of select

In prepration for configuring I2C/SPI defaults at the board level,
change the DesignWare GPIO driver's clock gate option to depend on clock
control instead of selecting clock control. This breaks a Kconfig
dependency loop when adding the following to a board's
Kconfig.defconfig:

config SPI
	default y if SENSOR

There aren't any in-tree users that enable the clock gate option, so
there aren't any places in-tree that now need to enable the clock
control driver. Out-of-tree users that set CONFIG_GPIO_DW_CLOCK_GATE=y
will now also need to set CONFIG_CLOCK_CONTROL=y.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>
